README.md
magento-ng-cookbook
===================

A collection of recipes to prepare environments for hosting Magento applications.

## Supported Platforms

TBC...

Recipes
-------

### magento-ng::cron

Sets up the system cron to call the Magento cron executor for each defined Magento site.

### magento-ng::etc-local

Prepared the local.xml file for each defined Magento site.

### magento-ng::stack

Set up a complete Magento application server calling many of the other magento-ng recipes.

## Attributes

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['magento']['db']['host']</tt></td>
    <td>String</td>
    <td>Database host name / IP</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['database']</tt></td>
    <td>String</td>
    <td>Database name</td>
    <td><tt>magentodb</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['username']</tt></td>
    <td>String</td>
    <td>Database user name</td>
    <td><tt>magentouser</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['password']</tt></td>
    <td>String</td>
    <td>Database user password</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['persistant']</tt></td>
    <td>String</td>
    <td>Use persistent connections</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['active']</tt></td>
    <td>String</td>
    <td>Is the database active</td>
    <td><tt>1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['model']</tt></td>
    <td>String</td>
    <td>Database type</td>
    <td><tt>mysql4</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['initStatements']</tt></td>
    <td>String</td>
    <td>Database init statements</td>
    <td><tt>SET NAMES utf8</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['db']['type']</tt></td>
    <td>String</td>
    <td>PHP database driver</td>
    <td><tt>pdo_mysql</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['base_path']</tt></td>
    <td>String</td>
    <td>Path to the root of the Magento files</td>
    <td><tt>public</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['locale']</tt></td>
    <td>String</td>
    <td>Default locale</td>
    <td><tt>en_GB</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['timezone']</tt></td>
    <td>String</td>
    <td>Default timezone</td>
    <td><tt>Europe/London</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['currency']</tt></td>
    <td>String</td>
    <td>Default currency</td>
    <td><tt>GBP</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['session_save']</tt></td>
    <td>String</td>
    <td>Session storage type (files|db|memcache)</td>
    <td><tt>db</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['admin_frontname']</tt></td>
    <td>String</td>
    <td>URL path to the admin</td>
    <td><tt>admin</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['use_rewrites']</tt></td>
    <td>String</td>
    <td>Whether to use URL rewriting</td>
    <td><tt>yes</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['use_secure']</tt></td>
    <td>String</td>
    <td>Allow secure connections</td>
    <td><tt>yes</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['use_secure_admin']</tt></td>
    <td>String</td>
    <td>Allow secure connections to admin</td>
    <td><tt>yes</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['multi_session_save']</tt></td>
    <td>String</td>
    <td>Multi session storage type (files|db|memcache)</td>
    <td><tt>db</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['session_memcache_ip']</tt></td>
    <td>String</td>
    <td>IP of session storage memcache instance</td>
    <td><tt>127.0.0.1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['session_memcache_port']</tt></td>
    <td>String</td>
    <td>Port to connect to session storage memcache</td>
    <td><tt>11211</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['backend_cache']</tt></td>
    <td>String</td>
    <td>Cache storage type (apc|memcache|xcache|file|CM_Cache_Backend_Redis)</td>
    <td><tt>file</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['app']['slow_backend']</tt></td>
    <td>String</td>
    <td>Slow cache storage type (database|file)</td>
    <td><tt>database</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['host']</tt></td>
    <td>String</td>
    <td>Hostnane/IP of the Redis instance</td>
    <td><tt>127.0.0.1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['port']</tt></td>
    <td>String</td>
    <td>Redis instance port</td>
    <td><tt>6379</tt></td>
  </tr>
    <tr>
      <td><tt>['magento']['redis']['password']</tt></td>
      <td>String</td>
      <td>Redis instance password</td>
      <td><tt></tt></td>
    </tr>
  <tr>
    <td><tt>['magento']['redis']['timeout']</tt></td>
    <td>String</td>
    <td>Redis connection timeout</td>
    <td><tt>2.5</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['database']</tt></td>
    <td>String</td>
    <td>Database to use for backend cache</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['full_page_cache_database']</tt></td>
    <td>String</td>
    <td>Database to use for full page cache</td>
    <td><tt>1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['session_database']</tt></td>
    <td>String</td>
    <td>Database to use for session storage</td>
    <td><tt>2</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['persistent']</tt></td>
    <td>String</td>
    <td>Unique string for persistent connections</td>
    <td><tt>c</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['full_page_cache_persistent']</tt></td>
    <td>String</td>
    <td>Unique string for persistent full page cache connections</td>
    <td><tt>fpc</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['session_persistent']</tt></td>
    <td>String</td>
    <td>Unique string for persistent session connections</td>
    <td><tt>s</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['force_standalone']</tt></td>
    <td>String</td>
    <td>Enforce standalone PHP redis, 0 for phpredis</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['connect_retries']</tt></td>
    <td>String</td>
    <td>Reduces errors due to random connection failures</td>
    <td><tt>1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['automatic_cleaning_factor']</tt></td>
    <td>String</td>
    <td>Enable automatic cleaning (not recommended)</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['compress_data']</tt></td>
    <td>String</td>
    <td>Enable data compression</td>
    <td><tt>1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['compress_tags']</tt></td>
    <td>String</td>
    <td>Enable tag compression</td>
    <td><tt>1</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['compress_threshold']</tt></td>
    <td>String</td>
    <td>Minimum string size for compression</td>
    <td><tt>2040</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['compression_lib']</tt></td>
    <td>String</td>
    <td>Compression library (gzip|lzf|l4z|snappy)</td>
    <td><tt>gzip</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['lifetime_limit']</tt></td>
    <td>String</td>
    <td>Lifetime for cache record</td>
    <td><tt>57600</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['log_level']</tt></td>
    <td>String</td>
    <td>Logging level</td>
    <td><tt>4</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['max_concurrency']</tt></td>
    <td>String</td>
    <td>Maximum number of processes that can wait for a lock</td>
    <td><tt>6</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['break_after_frontend']</tt></td>
    <td>String</td>
    <td>Seconds to wait for a session lock</td>
    <td><tt>5</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['fail_after']</tt></td>
    <td>String</td>
    <td>Seconds to wait before bailing (in addition to break after time)</td>
    <td><tt>10</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['break_after_adminhtml']</tt></td>
    <td>String</td>
    <td>Seconds to wait for a session lock in admin</td>
    <td><tt>30</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['first_lifetime']</tt></td>
    <td>String</td>
    <td>Lifetime of session for non-bots</td>
    <td><tt>600</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['bot_first_lifetime']</tt></td>
    <td>String</td>
    <td>Lifetime of session for bots on first write</td>
    <td><tt>60</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['bot_lifetime']</tt></td>
    <td>String</td>
    <td>Lifetime of session for bots on subsequent writes</td>
    <td><tt>7200</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['disable_locking']</tt></td>
    <td>String</td>
    <td>Disable session locking entirely</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['min_lifetime']</tt></td>
    <td>String</td>
    <td>Minimum session lifetime</td>
    <td><tt>60</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['max_lifetime']</tt></td>
    <td>String</td>
    <td>Maximim session lifetime</td>
    <td><tt>2592000</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['redis']['use_lua']</tt></td>
    <td>String</td>
    <td>Enable Lua for some operations</td>
    <td><tt>0</tt></td>
  </tr>
  <tr>
    <td><tt>['magento']['global']['extra_params']</tt></td>
    <td>Object</td>
    <td>Key/value parameter pairs</td>
    <td><tt>
      {
        'skip_process_modules_updates' => 1,
        'skip_process_modules_updates_dev_mode' => 1
      }
    </tt></td>
  </tr>
</table>

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

Supermarket share
-----------------

[stove](http://sethvargo.github.io/stove/) is used to create git tags and
publish the cookbook on supermarket.chef.io.

To tag/publish you need to be a contributor to the cookbook on Supermarket and
run:

```
$ stove login --username <your username> --key ~/.chef/<your username>.pem
$ rake publish
```

It will take the version defined in metadata.rb, create a tag, and push the
cookbook to http://supermarket.chef.io/cookbooks/magento-ng


License and Authors
-------------------
- Author:: Andy Thompson

```text
Copyright:: 2015 Inviqa UK LTD

See LICENSE file
```


CHANGELOG.md
## 2.0.0 (16 April 2018)

CHANGES:

  * #41 Bring in config-driven-helper v3.0.0

## 1.0.4 (1 November 2017)

BUG FIXES:

  * #39 Fix double-escaping the regular expression to deny access to RSS feeds.
  * #40 Rubocop style issues

## 1.0.3 (21 June 2017)

BUG FIXES:

  * #38 Improve RSS feed blocking on Apache HTTPD

## 1.0.2 (13 June 2017)

BUG FIXES:

  * #37 Block admin RSS feeds within store code prefixed urls

## 1.0.1 (23 May 2017)

BUG FIXES:

  * #36 Block admin RSS feeds

## 1.0.0 (27 October 2016)

CHANGES:

  * #32 Add additional redis related settings, with sensible defaults.

## 0.5.1 (14 September 2016)

BUG FIXES:

  * #34 Remove ignored logrotate missingok option

## 0.5.0 (06 May 2016)

CHANGES:

  * #31 Update config-driven-helper to allow for version >= 1.5, < 3.0

## 0.4.2 (03 May 2016)

CHANGES:

  * #25 Abbreviate copyright Limited to LTD for consistency
  * #28 Update code style and spec dependencies

BUG FIXES:

  * #29 Fix an incorrectly named block parameter in etc-local

## 0.4.1 (28 January 2016)

IMPROVEMENTS

  * Add additional new and existing restrictions from .htaccess

## 0.4.0 (10 November 2015)

FEATURES

  * Add a logrotate recipe to the stack for magento logs

BUG FIXES

  * Restrict access to sensitive files in root using new locations syntax

## 0.3.1 (02 October 2015)

FEATURES

  * Add support for include/exclude groups/jobs in AOE Scheduler cron

## 0.3.0 (21 September 2015)

FEATURES

  * Add support for AOE Scheduler cron and watchdog

## 0.2.1 (9 July 2015)

BUG FIXES

 * Fix crypt key attribute path

## 0.2.0 (3 July 2015)

IMPROVEMENTS

 * Reduce duplication in database configuration
 * Reduce restirctions on dependencies

## 0.1.1 (1 July 2015)

FEATURES

 * Add stack recipe

BUG FIXES

 * Fix app attribute path
 * Use raise in place of fatal exceptions

## 0.1.0 (7 June 2015)

FEATURES

  * Initial release of magento-ng


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

unless defined? Gem::MissingSpecError
  module Gem
    class MissingSpecError
    end
  end
end


spec/recipes/cron_spec.rb
describe 'magento-ng::cron' do
  context 'with default attributes and no magento sites' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.set['nginx']['sites']['project1']['type'] = 'non-magento-site'
        node.set['apache']['sites']['project2']['type'] = 'non-magento-site'
      end.converge(described_recipe)
    end

    it 'does not create cron.d file' do
      expect(chef_run).not_to create_cron_d('magento-project1')
      expect(chef_run).not_to create_cron_d('magento-project2')
    end
  end

  context 'with default attributes and one apache project site' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.set['nginx']['sites'] = {}
        node.set['apache']['sites']['project']['type'] = 'magento'
        node.set['apache']['sites']['project']['docroot'] = '/var/www/project/current/public'
      end.converge(described_recipe)
    end

    it 'creates project cron.d file' do
      expect(chef_run).to create_cron_d('magento-project').with(
        command: 'sh /var/www/project/current/public/cron.sh',
        minute: '*',
        user: 'www-data'
      )
    end
  end

  context 'with aoe schedule and watchdog' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.set['nginx']['sites'] = {}
        node.set['apache']['sites']['project']['type'] = 'magento'
        node.set['apache']['sites']['project']['docroot'] = '/var/www/project/current/public'
        node.set['apache']['sites']['project']['magento']['cron_type'] = 'aoe_scheduler'
      end.converge(described_recipe)
    end

    it 'creates project cron.d file using aoe schdule and watchdog' do
      expect(chef_run).not_to create_cron_d('magento-project')
      expect(chef_run).to create_template('/etc/cron.d/magento-project')
    end
  end

  context 'with default attributes and one nginx project site' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.set['apache']['sites'] = {}
        node.set['nginx']['sites']['project']['type'] = 'magento'
        node.set['nginx']['sites']['project']['docroot'] = '/var/www/project/current/public'
      end.converge(described_recipe)
    end

    it 'creates project cron.d file' do
      expect(chef_run).to create_cron_d('magento-project').with(
        command: 'sh /var/www/project/current/public/cron.sh',
        user: 'www-data'
      )
    end
  end
end


spec/recipes/stack_spec.rb
describe 'magento-ng::stack' do
  let(:chef_run) do
    ChefSpec::SoloRunner.new.converge(described_recipe)
  end

  it 'should include the cron recipe' do
    expect(chef_run).to include_recipe('magento-ng::cron')
  end

  it 'should include the etc-local recipe' do
    expect(chef_run).to include_recipe('magento-ng::etc-local')
  end
end


recipes/cron.rb
#
# Cookbook Name:: magento-ng
# Recipe:: cron
#
# Copyright (C) 2015 Inviqa UK LTD
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

%w[apache nginx].each do |type|
  node[type]['sites'].each do |name, site|
    next unless site['type'] == 'magento'

    include_recipe 'cron::default'

    primary_indicator_check = ''
    primary_indicator_check = "bash -c '[ -f #{site['clustered']['primary_indicator']} ] && " if site['clustered']

    magento = ConfigDrivenHelper::Util.immutablemash_to_hash(node['magento'])

    if site['magento']
      magento = ::Chef::Mixin::DeepMerge.hash_only_merge(
        magento,
        ConfigDrivenHelper::Util.immutablemash_to_hash(site['magento'])
      )
    end

    cron_user = if !site['cron'].nil? && site['cron']['user']
                  site['cron']['user']
                else
                  case type
                  when 'apache'
                    node['apache']['user']
                  when 'nginx'
                    node['php-fpm']['pools']['www']['user'].nil? ? node['php-fpm']['user'] : node['php-fpm']['pools']['www']['user']
                  end
                end

    if magento['cron_type'] == 'aoe_scheduler'
      template "/etc/cron.d/magento-#{name}" do
        source 'aoe-cron.erb'
        variables(
          :crons => magento['aoe_scheduler'],
          :cron_user => cron_user,
          :primary_indicator_check => primary_indicator_check,
          :site => site
        )
      end
    else
      cron_d "magento-#{name}" do
        command "#{primary_indicator_check}sh #{site['docroot']}/cron.sh"
        minute magento['cron']['minute']
        user cron_user
      end
    end
  end
end


recipes/etc-local.rb
#
# Cookbook Name:: magento-ng
# Recipe:: etc-local
#
# Copyright (C) 2015 Inviqa UK LTD
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

%w[apache nginx].each do |type|
  node[type]['sites'].each_pair do |name, site|
    next unless site['type'] == 'magento'

    magento = ConfigDrivenHelper::Util.immutablemash_to_hash(node['magento'])

    if site['magento']
      magento = ::Chef::Mixin::DeepMerge.hash_only_merge(
        magento,
        ConfigDrivenHelper::Util.immutablemash_to_hash(site['magento'])
      )
    end

    if Chef::Config[:solo]
      missing_attrs = %w[
        crypt_key
      ].select { |attribute| magento['app'][attribute].nil? }.map do |attribute|
        "node['#{type}']['sites']['#{name}']['magento']['app']['#{attribute}']"
      end

      raise "You must set #{missing_attrs.join(', ')} in chef-solo mode." unless missing_attrs.empty?
    else
      # generate all passwords
      %w[
        crypt_key
      ].select { |attribute| magento['app'][attribute].nil? }.map do |attribute|
        node.set_unless[type]['sites'][name]['magento']['app'][attribute] = secure_password
        node.save
      end
    end

    config_path = if site['capistrano']
                    "#{site['capistrano']['deploy_to']}/shared/#{magento['app']['base_path']}"
                  else
                    site['docroot']
                  end

    directory "#{config_path}/app/etc" do
      recursive true
      action :create
      not_if do
        ::File.exist?("#{config_path}/app/etc")
      end
    end

    template "#{config_path}/app/etc/local.xml" do
      source 'magento-local.xml.erb'
      mode 0o0644
      variables(
        :magento => magento
      )
    end
  end
end


recipes/logrotate.rb
#
# Cookbook Name:: magento-ng
# Recipe:: logrotate
#
# Copyright (C) 2015 Inviqa UK LTD
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

%w[apache nginx].each do |type|
  node[type]['sites'].each_pair do |name, site|
    next unless site['type'] == 'magento'

    magento = ConfigDrivenHelper::Util.immutablemash_to_hash(node['magento'])

    if site['magento']
      magento = ::Chef::Mixin::DeepMerge.hash_only_merge(
        magento,
        ConfigDrivenHelper::Util.immutablemash_to_hash(site['magento'])
      )
    end

    config_path = if site['capistrano']
                    "#{site['capistrano']['deploy_to']}/shared/#{magento['app']['base_path']}"
                  else
                    site['docroot']
                  end

    logrotate_app "magento-#{name}" do
      path "#{config_path}/var/log/*.log"
      unless magento['logrotate'].nil?
        magento['logrotate'].each do |key, value|
          send key, value
        end
      end
    end
  end
end


attributes/default.rb
default['magento']['db']['host'] = 'localhost'
default['magento']['db']['database'] = 'magentodb'
default['magento']['db']['username'] = 'magentouser'
default['magento']['db']['persistent'] = '0'
default['magento']['db']['active'] = '1'
default['magento']['db']['model'] = 'mysql4'
default['magento']['db']['initStatements'] = 'SET NAMES utf8'
default['magento']['db']['type'] = 'pdo_mysql'

default['magento']['db_connections']['default_setup'] = {}

default['magento']['app']['base_path'] = 'public'
default['magento']['app']['locale'] = 'en_GB'
default['magento']['app']['timezone'] = 'Europe/London'
default['magento']['app']['currency'] = 'GBP'
default['magento']['app']['session_save'] = 'db' # files|db|memcache
default['magento']['app']['admin_frontname'] = 'admin'
default['magento']['app']['use_rewrites'] = 'yes'
default['magento']['app']['use_secure'] = 'yes'
default['magento']['app']['use_secure_admin'] = 'yes'
default['magento']['app']['multi_session_save'] = 'db' # files|db|memcache|redis
default['magento']['app']['session_memcache_ip'] = '127.0.0.1'
default['magento']['app']['session_memcache_port'] = '11211'
default['magento']['app']['backend_cache'] = 'file' # apc|memcached|xcache|file|Cm_Cache_Backend_Redis
default['magento']['app']['slow_backend'] = 'database' # database|file
default['magento']['app']['backend_servers'] = []

default['magento']['redis']['host'] = '127.0.0.1'
default['magento']['redis']['port'] = '6379'
default['magento']['redis']['password'] = ''
default['magento']['redis']['timeout'] = '2.5'
default['magento']['redis']['database'] = '0'
default['magento']['redis']['full_page_cache_database'] = '1'
default['magento']['redis']['session_database'] = '2'
default['magento']['redis']['persistent'] = 'c'
default['magento']['redis']['full_page_cache_persistent'] = 'fpc'
default['magento']['redis']['session_persistent'] = 's'
default['magento']['redis']['force_standalone'] = '0'
default['magento']['redis']['connect_retries'] = '1'
default['magento']['redis']['automatic_cleaning_factor'] = '0'
default['magento']['redis']['compress_data'] = '1'
default['magento']['redis']['compress_tags'] = '1'
default['magento']['redis']['compress_threshold'] = '2048'
default['magento']['redis']['compression_lib'] = 'gzip'
default['magento']['redis']['lifetime_limit'] = '57600'
default['magento']['redis']['log_level'] = '4'
default['magento']['redis']['max_concurrency'] = '6'
default['magento']['redis']['break_after_frontend'] = '5'
default['magento']['redis']['fail_after'] = '10'
default['magento']['redis']['break_after_adminhtml'] = '30'
default['magento']['redis']['first_lifetime'] = '600'
default['magento']['redis']['bot_first_lifetime'] = '60'
default['magento']['redis']['bot_lifetime'] = '7200'
default['magento']['redis']['disable_locking'] = '0'
default['magento']['redis']['min_lifetime'] = '60'
default['magento']['redis']['max_lifetime'] = '2592000'
default['magento']['redis']['use_lua'] = '0'

default['magento']['global']['custom'] = ''
default['magento']['global']['extra_params'] = {
  'skip_process_modules_updates' => 1,
  'skip_process_modules_updates_dev_mode' => 1
}

default['magento']['cron_type'] = 'standard'
default['magento']['cron']['minute'] = '*'
default['magento']['aoe_scheduler']['always']['mode'] = 'always'
default['magento']['aoe_scheduler']['always']['minute'] = '*'
default['magento']['aoe_scheduler']['default']['mode'] = 'default'
default['magento']['aoe_scheduler']['default']['minute'] = '*'
default['magento']['aoe_scheduler']['watchdog']['mode'] = 'watchdog'
default['magento']['aoe_scheduler']['watchdog']['minute'] = '*/10'

default['magento']['logrotate']['frequency'] = 'weekly'
default['magento']['logrotate']['rotate'] = 4

%w[apache nginx].each do |type|
  default[type]['shared_config']['magento'] = {
    'clustered' => false,
    'type' => 'magento',
    'protocols' => %w[http https],
    'locations' => {
      '/app' => 'restricted',
      '/dev' => 'restricted',
      '/downloader' => 'restricted',
      '^/errors/.*\\.(xml|phtml)$' => {
        'type' => 'regex',
        'mode' => 'restricted'
      },
      '/includes' => 'restricted',
      '/lib' => 'restricted',
      '/media' => 'static',
      '/media/custom_options' => 'restricted',
      '/media/customer' => 'restricted',
      '/media/downloadable' => 'restricted',
      '/pkginfo' => 'restricted',
      '/report/config.xml' => 'restricted',
      '^(/index\.php)?/?(.+/)?rss/(catalog/(notifystock|review)|order/new)(/|$)' => {
        'type' => 'regex',
        'mode' => 'restricted'
      },
      '/shell' => 'restricted',
      '/skin' => 'static',
      '/var' => 'restricted',
      '/.gitignore' => 'restricted',
      '/composer.json' => 'restricted',
      '/cron.php' => 'restricted',
      '/cron.sh' => 'restricted',
      '/index.php.sample' => 'restricted',
      '/install.php' => 'restricted',
      '/mage' => 'restricted',
      '/php.ini.sample' => 'restricted',
      '/RELEASE_NOTES.txt' => 'restricted',

      '/\\.ht[^/]*$' => {
        'type' => 'regex',
        'mode' => 'restricted'
      }
    },
    'capistrano' => {
      'owner' => 'deploy',
      'group' => 'deploy',
      'shared_folders' => {
        'standard' => {
          'folders' => [
            'public/app/etc/'
          ]
        },
        'writeable' => {
          'owner' => 'apache',
          'group' => 'apache',
          'folders' => [
            'public/./media',
            'public/./sitemaps',
            'public/./staging',
            'public/./var'
          ]
        }
      }
    }
  }
end
